### ISSUE TITLE
Add AuthMetadataProcessor support to InsecureServerCredentials

### ISSUE THREAD

**Issue by BurningEnlightenment** [grpc/grpc#21589]
### Is your feature request related to a problem? Please describe.
I host multiple grpc services and http apps on a single machine behind an NGINX reverse proxy [[1]](https://www.nginx.com/blog/nginx-1-13-10-grpc/) which forwards incoming requests to the correct service based on TLS SNI. I.e. the client connects via TLS to the publicly exposed NGINX server.

I also rely on a third party authentication solution which I integrate in grpc via the `AuthMetadataProcessor`. However, I can only use the custom `AuthMetadataProcessor` with `SslServerCredentials` which implies that I'm currently forced to use TLS for the NGINX <=> grpc service connection.

It is pretty annoying to provision the grpc services with self signed certificates for the loopback address. Furthermore it is an outright waste of server resources to encrypt connections on the loopback adapter.

### Describe the solution you'd like
```cpp
std::shared_ptr<AuthMetadataProcessor> authProcessor  = /*...*/;
auto serverCredentials = grpc::InsecureServerCredentials();
serverCredentials->SetAuthMetadataProcessor(std::move(authPocessor));
```
To work the same way as it would with `SslServerCredentials()`.

### Describe alternatives you've considered
I wouldn't mind if another `ServerCredentials` implementation (maybe `MetadataServerCredentials`) would be provided in order to keep the completely unauthenticated case separated.

### Additional context

[1] https://www.nginx.com/blog/nginx-1-13-10-grpc/
[2] related #17942
[3] similar use case https://groups.google.com/forum/#!topic/grpc-io/ifL63H0kN48

**Comment by veblush** [comment]
Jiangtao, can you take a look at this?

**Comment by sazconaitbit** [comment]
im looking for the same, any news about this? thanks

**Comment by veblush** [comment]
@yihuazhang Could you take a look at this?

**Comment by toddtomkinson** [comment]
The current state of #25586 will NOT allow this, as the SetAuthMetadataProcessor in InsecureServerCredentials still does not allow an AuthMetadataProcessor to be set:

```
  void SetAuthMetadataProcessor(
      const std::shared_ptr<grpc::AuthMetadataProcessor>& processor) override {
    (void)processor;
    GPR_ASSERT(0);  // Should not be called on InsecureServerCredentials.
  }
```

It appears that the default implementation will only be used by classes that extend SecureServerCredentials.

**Comment by markdroth** [comment]
Sorry, I somehow missed that last comment.

Now that #25586 has been merged, I agree that we should change insecure creds to allow specifying an AuthMetadataProcessor.

Zhen, can you please take a look at this?  Thanks!

**Comment by ZhenLian** [comment]
Yeah, this looks reasonable, and it seems we just need to replace the original `SetAuthMetadataProcessor` code with the code in https://github.com/grpc/grpc/blob/master/src/cpp/server/secure_server_credentials.cc#L101-L104.
I will be OOO until next week. @toddtomkinson , interested in submitting a PR to quickly fix this:)?

**Comment by markdroth** [comment]
I think we probably want to just eliminate `SecureServerCredentials` completely and instead move its functionality directly into `ServerCredentials`, so that there is no difference between insecure creds and other creds types.

**Comment by ZhenLian** [comment]
Yeah, I agree. There seems to be some clean up work for `ServerCredentials` as well, and we might want to do that as well.

**Comment by dywalsh-arista** [comment]
Curious to know if there has been any updates on this? Thanks

**Comment by ZhenLian** [comment]
@erm-g Would you mind taking a look at this please? Thank you!

**Comment by olafmustafar** [comment]
Hi, any progress on this?

**Comment by chihong0522** [comment]
HI, want the same feature to auth without SSL

**Comment by kzolti** [comment]
Hi, I want to use AuthMetadataProcessor for authentication without SSL, any progress on this?
`insecure_server_credentials.cc:42]    ASSERTION FAILED: 0`
gRPC version: 37.0.0

**Comment by huerni** [comment]
Hi, Has the issue been resolved?

**Comment by rchaikozova** [comment]
Hi, any update on this?
